### ISSUE TITLE
here() function doesn't work, but setwd() does with the exact same file path.

### ISSUE THREAD

**Issue by don-radcliffe** [don-radcliffe/nps_fx#1]
@mdscheuerell I was hoping you could help me with a weird issue with here() that I've been having.  When I run it at the beginning of a script it sets the correct working directory, but then it can't open the file connection.  If I set the exact same working directory with setwd(), it works fine.  Furthermore, if I run the exact same here() call *after* running the setwd() command, it opens my file just fine.   I downloaded the newest version of R and I'm still having the problem.

I'm not able to make a fully contained reprex or do this within RMarkdown, because it's an issue with reading the data and RMarkdown does working directories differently than normal scripts.  So what I've done to make a reprex is pasted my code from a normal R script below; you could also pull the 'here_reprex.R' file from my 'scripts/data_processing' folder in this repo.   Single hashtags denote R's output in response to a command, double hashtags are my writing.  If you wanted to make this a full reprex, you could clone this repo.  The link is https://github.com/don-radcliffe/nps_fx.  

No worries if you don't see a quick solution to this; it's a very minor issue at the moment and I'll put it on Stack Overflow if there isn't something obvious that I'm missing.   Thanks so much!

My R script starts below:

## First try of here() doesn't open file connection

library(here)
#here() starts at C:/ProgramR/nps_fx

here('data')
#[1] "C:/ProgramR/nps_fx/data"

trees <- read.csv('data_raw/trees_raw.csv')
#Error in file(file, "rt") : cannot open the connection
#In addition: Warning message:
  #In file(file, "rt") :
  #cannot open file 'data_raw/trees_raw.csv': No such file or directory

## setwd() successfully opens file connection with the exact same path

##Sorry Jenny Bryan
setwd('C:/ProgramR/nps_fx/data')

trees <- read.csv('data_raw/trees_raw.csv')

head(trees, 2)
#Date Monitoring.Status MacroPlot.Name MacroPlot.UTM.Zone MacroPlot.UTM.Y MacroPlot.UTM.X
#1 5/29/1996 0:00             00Pre   FPSME2D08-21                11U         5357384          668304
#2 5/29/1996 0:00             00Pre   FPSME2D08-21                11U         5357384          668304

## here() successfully open file connection after the setwd() command is called

here('data')

trees <- read.csv('data_raw/trees_raw.csv')

head(trees, 2)
#Date Monitoring.Status MacroPlot.Name MacroPlot.UTM.Zone MacroPlot.UTM.Y MacroPlot.UTM.X
#1 5/29/1996 0:00             00Pre   FPSME2D08-21                11U         5357384          668304
#2 5/29/1996 0:00             00Pre   FPSME2D08-21                11U         5357384          668304

##  Session Information

sessionInfo()
#R version 4.0.5 (2021-03-31)
#Platform: x86_64-w64-mingw32/x64 (64-bit)
#Running under: Windows 10 x64 (build 19041)

#Matrix products: default

#locale:
#[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
#[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
#[5] LC_TIME=English_United States.1252    

#attached base packages:
#[1] stats     graphics  grDevices utils     datasets  methods   base     

#other attached packages:
#[1] here_1.0.1

#loaded via a namespace (and not attached):
#[1] compiler_4.0.5  rprojroot_2.0.2 tools_4.0.5

**Comment by mdscheuerell** [comment]
Hi Don.

OK, this looks like an easy fix. In this case the data you'd like to read in are 2 folders deep from the main working directory. That is,

`/data/data_raw`

so I'd recommend you start with something like

```
## use here() with 2+ names of folders/dirs
raw_data_dir <- here::here("data, "data_raw")
tidy_data_dir <- here::here("data, "data_tidy")
```

I'd also recommend the `file.path()` function as a defensive way to create proper file paths, so something like

```
## read the raw data
trees <- read.csv(file.path(raw_data_dir, "trees_raw.csv"))
```

You should then be able to write the tidy data with

```
## assumes tidy data are in `trees_tidy`
write.csv(trees_tidy, file.path(tidy_data_dir, "trees_tidy.csv"))
```

**Comment by don-radcliffe** [comment]
Thanks @mdscheuerell!  I was about to tell you nevermind on this issue; I dug deeper into here() and now I see that I've been misunderstanding how calling here() works under the hood.   It worked for me in a different repo because I happened to have the file structure set up in a certain way.  Once I realized this I was able to make it work by calling:

```{r here example}
trees <- read.csv(here('data/data_raw/trees_raw.csv'))
```
But I'm glad you responded so quickly, because you've given me some good new defensive ways to read and write data, that are ultimately less clunky and more reliable than what I did above.  This stuff is going to be super useful on the collaborative projects I'm starting; thanks as always for your help!

Don
